### PR TITLE
Python external data conversion - Add support for aligning data for mmap

### PR DESCRIPTION
**Pull request by ivberg**
### Description
Add optional (default off) support for Python external data conversion aligning offset necessary for memory mapping support

Note, also fixes 2 bugs that were previously in the code
1. Length was being calculated incorrectly, so that the tensor size filter might not work as a user expects
2. If the output file was already present, the file was not cleared, and thus the zero paddings were not being written

### Motivation and Context
Without this memory mapping does not actually work
#6226 
#6247
Also see https://github.com/microsoft/onnxruntime/issues/21195

### CODE REVIEW

File: onnx/__init__.py [hunk 1]
```diff
@@ -287,6 +287,9 @@ def save_model(
     location: str | None = None,
     size_threshold: int = 1024,
     convert_attribute: bool = False,
+    align_offset: bool = False,
```
Comment (justinchuby) [inline onnx/__init__.py:290]: Should we default this to true?

File: onnx/external_data_helper.py [hunk 2]
```diff
@@ -134,17 +141,25 @@ def convert_model_to_external_data(
...
             if (
                 tensor.HasField("raw_data")
-                and sys.getsizeof(tensor.raw_data) >= size_threshold
+                and tensor_size >= size_threshold
             ):
-                set_external_data(tensor, file_name)
+                if align_offset and tensor_size > align_threshold:
+                    # Align to the larger of the page size or the allocation granularity
+                    alignment_factor = max(4096, allocation_granularity)  # Assuming 4k page size
+                    current_offset = (current_offset + alignment_factor - 1) // alignment_factor * alignment_factor # Align to the next page or alloc granularity boundary
```
Comment (justinchuby) [inline onnx/external_data_helper.py:154]: Possible to extract this logic into a private function and create a test for it?

Comment (ivberg) [reply onnx/external_data_helper.py:154]: Agree we should have a test. However, I am not sure I have the bandwidth. Is this something that someone from ONNX could sponsor / contribute code to @justinchuby. yourself or someone else?

Note that @pranavsharma suggested other changes. Such as passing in the OS "Windows, UNIX, MacOsX" as string parameter and determining page size and VirtualAlloc memory allocation granularity based on the OS string

Comment (justinchuby) [reply onnx/external_data_helper.py:154]: Ideally to me the model file should be os agnostic. Picking the least common multiple in this case as default makes sense to me

File: onnx/__init__.py (hunk 1 above)
Comment (ivberg) [reply onnx/__init__.py:290]: The argument made against this was to keep consistency? This is a behavior change IMO, but not really a breaking behavior change. There is also an increase in model file size with the padding added.

I am in favor of eventually defaulting this on, but I also think ORT needs better support of mapfiles first to save a lot commit memory cost?
See https://github.com/microsoft/onnxruntime/issues/21448

File: onnx/external_data_helper.py [hunk 3]
```diff
@@ -298,12 +308,21 @@ def write_external_data_tensors(model: ModelProto, filepath: str) -> ModelProto:
...
         # 2. The raw data in these tensors is serialized to a file
         # Thus serialize only if tensor has raw data and it was marked for serialization
         if uses_external_data(tensor) and tensor.HasField("raw_data"):
+            if not file_cleared:
+                # Clear the file if it exists, then open for reading and writing
+                # Clear needed for old data, to ensure file_size grows and padding is added
+                external_data_file_path = os.path.join(filepath, ExternalDataInfo(tensor).location)
+                with open(external_data_file_path, "wb"):
+                    pass
+                file_cleared = True
```
Comment (gramalingam) [inline onnx/external_data_helper.py:325]: (a) Seems to me this may not work as intended when we write each tensor to a separate file?
(b) Also, I guess this is an orthogonal improvement (trying to handle pre-existing files), independent of the alignment extension?

File: onnx/external_data_helper.py [hunk 4]
```diff
@@ -134,17 +141,25 @@ def convert_model_to_external_data(
...
+                    current_offset = (current_offset + alignment_factor - 1) // alignment_factor * alignment_factor # Align to the next page or alloc granularity boundary
+
+                set_external_data(tensor, file_name, offset=current_offset)
+                current_offset += tensor_size
     else:
         for tensor in tensors:
             if (
                 tensor.HasField("raw_data")
-                and sys.getsizeof(tensor.raw_data) >= size_threshold
+                and len(tensor.raw_data) >= size_threshold
```
Comment (gramalingam) [inline onnx/external_data_helper.py:162]: This looks reasonable ... but I wonder why they used `sys.getsizeof` originally? Are the two guaranteed to return the same value in all cases?

File: onnx/external_data_helper.py (hunk 3 above)
Comment (ivberg) [reply onnx/external_data_helper.py:325]: It's not orthogonal in that I hit a bug when writing out the padding (zero's). It didn't work when there was a pre-existing file because the code uses the fact that the file size is increasing to know when to write zero's.

File: onnx/external_data_helper.py (hunk 4 above)
Comment (ivberg) [reply onnx/external_data_helper.py:162]: sys.getsizeof()  in Python apparently returns the size of the object, not the actual length of the tensor raw data. I kept hitting a bug with the previous code where it would write out too big of the data + padding vs what the size of the tensor was originally. This is because of this incorrect length calculation. I think nobody noticed the bug before because it was "close enough" to use as a size_threshold

Comment (ivberg) [reply onnx/external_data_helper.py:162]: for (a) I would expect this would work with multiple files but I admit I did not test that. Code review should tell us the answer

File: onnx/external_data_helper.py (hunk 3 above)
Comment (gramalingam) [reply onnx/external_data_helper.py:325]: How would it work for multiple files? If two of the files are pre-existing, file_cleared would be set to true after the first file, and so the clear logic will not be invoked for the second file.

File: onnx/check-model-extdata-alignment.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
+import os
```
Comment (gramalingam) [inline onnx/check-model-extdata-alignment.py:1]: Is this intended to be part of the PR? I don't understand its purpose. I am also concerned about the code for extracting zip files etc., which have triggered some vulnerability issues in the past. It may be better to avoid such utilities, unless they are important.

Comment (justinchuby) [reply onnx/check-model-extdata-alignment.py:1]: Just an example test script @ivberg explained to me. I can take over the PR and formalize the implementation.

File: onnx/__init__.py (hunk 1 above)
Comment (frank-dong-ms) [reply onnx/__init__.py:290]: I agree we default this to true, I'm working a similar offset alignment PR from ORT: https://github.com/microsoft/onnxruntime/pull/21604 
We want to enable offset alignment when save external data file during inference session initialization by default. With that, there are some optimization script/tool using onnx save_model method to save onnx model (for external initializer as well), for example https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/onnx_model.py#L1164, with align offset default to true in onnx and in ORT, we can have consistent behavior.

**Comment by justinchuby** [comment]
This feature is implemented by the PyTorch -> ONNX exporter and in ONNX Script as a utility. Users can leverage or replicate the logic for this purposes. Closing this for now to reduce complexity of this api. Thanks!
